"Customizing your Mappings" article (#86)
QuiltMC/developer-wiki: wiki/misc/mappings/en.md

development environment, unmapped things will use their intermediary name instead of the obfuscated one, so if you have
ever browsed Minecraft's code with QM or Fabric's Yarn applied, it's very likely you have seen quite a few intermediary
names. It gets worse with method parameters: since they are really prone to incompatible changes between versions, they
(usually) don't have intermediate names, thus the names you see in the code depend on the tooling that you used to
decompile the game.

If you want to add a name or change a wrong or bad one, or add documentation to the code, you can start by picking which
mapping set you'll work on top of. In this article, we'll use QM, though the process for Yarn is almost identical.
If you want to work on top of Mojang's mappings, you'll have to do some extra work which we won't cover here. If you
are going to work on QM, we highly suggest taking a look at its [contributing documentation][QM CONTRIBUTING.md] and
contributing your changes to the repository. You'll need some very basic Git knowledge, but it should be fairly easy
to do if you've ever worked with Git before.

> To get started, clone or download [Quilt Mappings] (if you want to contribute, [fork the repo][fork qm] and clone said
> fork), and run `./gradlew mappings` in your command prompt or terminal. This will launch [Enigma], our tool to edit and
> write mappings. Rai wrote a really [great guide on how to edit mappings in Enigma][Enigma guide], so you can take a look
> and start mapping! Once you have finished editing, don't forget to save your changes before closing Enigma.

Author: How's this instead?

```suggestion
To get started, first get your own copy of [Quilt Mappings]' code by cloning or downloading the repo. If you want to
contribute your changes at some point, first [fork the repo][fork qm] and clone said fork instead (directly downloading
the code won't work). Once you have the code, run `./gradlew mappings` in your command prompt or terminal, which will
launch [Enigma], our favorite tool to edit and write mappings.
Rai wrote a really [great guide on how to edit mappings in Enigma][Enigma guide], so you can take a look and start
mapping! Once you have finished editing, don't forget to save your changes before closing Enigma.
```

# Customizing the Mappings in Your Development Environment

## What are mappings?

> Before a new Minecraft version jar is published to Mojang's servers, it goes through a process called *obfuscation*, where the human-readable class, field and method names are simplified to just a few letters, mainly to optimize the file size, and in other contexts to prevent people from easily understanding your code. Obfuscation makes code very difficult to understand, because most names are just converted to random letters, so this is where mappings come into play.

Reviewer: There are two sentences talking about making code hard to understand. I would remove the clause from the first sentence and add a `In addition` or something to the second sentence. Probably also split up the last sentence. 

```suggestion
Before a new Minecraft version jar is published to Mojang's servers, it goes through a process called *obfuscation*, where the human-readable class, field and method names are simplified to just a few letters, mainly to optimize the file size. In addition, obfuscation makes code very difficult to understand, because most names are just converted to random letters. This is where mappings come into play.
```

A mapping is just a change from one name to another, in most cases an obfuscated name to a human readable one. Each mapping may also have extra metadata, like documentation. A set of mappings is called a "mapping set" or more often just "mappings". You can think of a mapping set as a translation dictionary, where each single mapping would be a translation of a word to another language. Here, the obfuscated names would be the language the computer uses, and the dictionary would help us translate to plain english.

> While a simple obuscated-to-human-readable mapping set would be enough for one minecraft version, the obfuscated names aren't constant between different minecraft versions: `DirtBlock` could be obfuscated as `abc` in 1.19 but in 1.20 it could be `def`. From this arises the need to keep the obfuscation changes between versions minimal, and the problem can be solved with intermediary mappings, that convert those obfuscated names to names that won't change between versions, but still aren't readable in english. Quilt uses Hashed Mojmap, in which every class, field, and method is prefixed by `C_`, `f_` and `m_` respectively, followed by an 8-letter hash. While developing mods, most of the you'll see Fabric's intermediary instead, which uses `class_`, `field_` and `method_`, followed by a number. Mojang publishes official mappings for every version 1.14, but they don't use intermediaries, so you have to do some processing to use them where you'd use other mapping sets.

Reviewer: `intermediary` to `intermediate`

> While a simple obuscated-to-human-readable mapping set would be enough for one minecraft version, the obfuscated names aren't constant between different minecraft versions: `DirtBlock` could be obfuscated as `abc` in 1.19 but in 1.20 it could be `def`. From this arises the need to keep the obfuscation changes between versions minimal, and the problem can be solved with intermediary mappings, that convert those obfuscated names to names that won't change between versions, but still aren't readable in english. Quilt uses Hashed Mojmap, in which every class, field, and method is prefixed by `C_`, `f_` and `m_` respectively, followed by an 8-letter hash. While developing mods, most of the you'll see Fabric's intermediary instead, which uses `class_`, `field_` and `method_`, followed by a number. Mojang publishes official mappings for every version 1.14, but they don't use intermediaries, so you have to do some processing to use them where you'd use other mapping sets.

Reviewer: Missing a time in the sentence starting with `When developing mods`

> While a simple obuscated-to-human-readable mapping set would be enough for one minecraft version, the obfuscated names aren't constant between different minecraft versions: `DirtBlock` could be obfuscated as `abc` in 1.19 but in 1.20 it could be `def`. From this arises the need to keep the obfuscation changes between versions minimal, and the problem can be solved with intermediary mappings, that convert those obfuscated names to names that won't change between versions, but still aren't readable in english. Quilt uses Hashed Mojmap, in which every class, field, and method is prefixed by `C_`, `f_` and `m_` respectively, followed by an 8-letter hash. While developing mods, most of the you'll see Fabric's intermediary instead, which uses `class_`, `field_` and `method_`, followed by a number. Mojang publishes official mappings for every version 1.14, but they don't use intermediaries, so you have to do some processing to use them where you'd use other mapping sets.

Reviewer: missing an `after` in `Mojang publishes...`

## Editing your mappings

Mapping sets don't have to contain a mapping for every class, field or method in a jar to be valid, and in fact most mapping sets aren't complete. For instance, Quilt Mappings (QM for short) has reached 99% completion at most. Unmapped things in a development environment will use their intermediary name, instead of the obfuscated one, so if you have ever browsed Minecraft's code with QM or Fabric's Yarn applied, you have probably seen quite a few intermediary names. It gets worse with parameters, since they are really prone to incompatible changes between versions, and they don't have intermediary names, so the names you see in the code depend on what tooling you used to decompile the game.

> If you want to add or change a wrong or bad name, or add documentation to the code, you can start by picking what mapping set you'll work on top of. In this article, we'll use QM, though the process for Yarn is almost identical. If you want to work on top of Mojang's mappings, you'll have to do some extra work which we won't cover here. If you are going to work with QM we highly suggest looking at its [CONTRIBUTING.md](https://github.com/QuiltMC/quilt-mappings/blob/HEAD/CONTRIBUTING.md) and contributing your changes to the repository.

Reviewer: Add a bit more on contributing to upstream.

# Customizing the Mappings in Your Development Environment

## What are mappings?

Before a new Minecraft version jar is published to Mojang's servers, it goes through a process called *obfuscation*,
where the human-readable class, field and method names are simplified to just a few letters, mainly to optimize the file

> size. In addition, obfuscation makes code very difficult to understand, because most names are replaced by a few random

Reviewer 2: ```suggestion
size. In addition, obfuscation makes code very difficult to understand, because those simplified names aren't just contractions of the real names, they're completely random
```
made the wording a little more fun

Author: Whoops

just "mappings". You can think of a mapping set as a translation dictionary, where each single mapping would be a
translation of a word to another language. Here, the obfuscated names would be the language the computer uses, and
the dictionary would help us translate that to plain english.

While a simple obuscated-to-human-readable mapping set would be enough for one minecraft version, the obfuscated names
aren't constant between different minecraft versions: `DirtBlock` could be obfuscated as `abc` in 1.19 but in 1.20 it
could be `def`. From this arises the need to keep the obfuscation changes between versions minimal, and the problem can
be solved with intermediate mappings, that convert those obfuscated names to names that won't change between versions,
but still aren't readable in english. Quilt uses Hashed Mojmap, in which every class, field, and method is prefixed by
`C_`, `f_` and `m_` respectively, followed by an 8-letter hash. While developing mods, most of the time you'll see
Fabric's intermediary instead, which uses `class_`, `field_` and `method_`, followed by a number. Mojang also publishes

> official mappings, often called Mojmap (short for Moj(ang)-map(pings)), for every version after 1.14, but, since they don't use
> intermediate mappings, you have to do some extra processing to use them where you'd use other mapping sets. Luckily, Loom

Reviewer: ```suggestion
official mappings, often called Mojmap (short for Moj(ang)-map(pings)), for every version after 1.14. Since they don't have an
intermediate mapping, you have to do some extra processing to use them where you'd use other mapping sets. Luckily, Loom
```

does this process for you, so you don't have to worry about that and can easily replace the mappings used in your mod for
Mojang's oficial ones.

There a few different formats to store mappings, and as of the time this article was written, we are developing a new

> one with a bunch of improvements. In our ecosystem, the most used format is Tiny V2, which uses a single `.tiny` file to

Reviewer: I would link to the fabric tinyv2 spec. i think they have moved it finally from that one issue on TR

store mappings, and places fields and methods as "children" of their parent class. Another format we often use is the

> Enigma format, which uses a directory tree with a file for each top-level class, and organize the entries in a

Reviewer 2: ```suggestion
Enigma format, which uses a directory tree with a file for each top-level class, and organizes the entries in a
```
proper tense here

ever browsed Minecraft's code with QM or Fabric's Yarn applied, it's very likely you have seen quite a few intermediary
names. It gets worse with method parameters: since they are really prone to incompatible changes between versions, they
(usually) don't have intermediate names, thus the names you see in the code depend on the tooling that you used to
decompile the game.

If you want to add or change a wrong or bad name, or add documentation to the code, you can start by picking which
mapping set you'll work on top of. In this article, we'll use QM, though the process for Yarn is almost identical.
If you want to work on top of Mojang's mappings, you'll have to do some extra work which we won't cover here. If you
are going to work on QM, we highly suggest taking a look at its [CONTRIBUTING.md](https://github.com/QuiltMC/quilt-mappings/blob/HEAD/CONTRIBUTING.md)
and contributing your changes to the repository. You'll need some very basic Git knowledge, but it should be fairly
easy if you've ever worked with Git before.

To get started, clone or download [Quilt Mappings](https://github.com/QuiltMC/quilt-mappings) (if you want to

> contribute, [fork the repo] and clone it), and run `./gradlew mappings` in your command prompt or terminal. This will

Reviewer 2: is there supposed to be a link on [fork the repo] ?

Author: Yep, forgot to add the link lol

> launch [Enigma](https://github.com/QuiltMC/enigma), our tool to edit and write mappings. Rai wrote a really

Reviewer 2: awh

[great guide on how to edit mappings in Enigma](https://github.com/QuiltMC/quilt-mappings/blob/HEAD/GUIDE.md), so you

> can take a look and start mapping!

Reviewer: ```suggestion
can take a look and start mapping! Once you have finished editing, don't forget to save your changes before closing Enigma.
```

> To contribute your changes, you have to add and commit your changes, and then push the changes to your fork of QM. This

Reviewer: ```suggestion
### Contributing the changes back to Quilt
To contribute your changes, you have to add and commit your changes, and then push the changes to your fork of QM. This
```

is really easy to do with an IDE, as described in [the Setting Up article](/en/introduction/setting-up), but you can
also do it from the command prompt or terminal with these commands.
```bash
git add . # tell git to track all your changes in the current directory

git commit -m "Blabla" # add those changes into a new commit (replace "blabla" with a short description of your changes)

git push # upload your changes to your fork of QM. You might need to add `origin <minecraft version>` at the end if git complains about a missing upstream branch
```

Once you pushed your changes to your fork, go to [QM's Pull Requests tab](https://github.com/QuiltMC/quilt-mappings/pulls),
and click the "Compare & Pull Request" button in the note about your recent changes. Fill in the title and description of
your PR, submit it, and wait for your changes to be reviewed and accepted. Again, there's a more in-depth explanation of
the PR process in [CONTRIBUTING.md](https://github.com/QuiltMC/quilt-mappings/blob/HEAD/CONTRIBUTING.md).

> ## Using the edited mappings
> 
> Once you have finished and saved your changes, you can run `./gradlew publishToMavenLocal` to make the required files

Reviewer: ```suggestion
### Using the edited mappings

If you don't want to contribute your changes back to Quilt, or want to try them out in a development environment, you can run `./gradlew publishToMavenLocal` to make the required files
```
